packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2680,46 +2676,80 @@ export class ContainerRuntime
 				local,
 			);
 			if (messagesWithPendingState.length > 0) {
```

Comment:
I think this code might read better if you pulled everything in this `if` block into a function `processBatch` (same signature as `processEmptyBatch`), and then in-lined `processRuntimeMessage` (or not, I guess).  It keeps this function more concise and focused on the overall flow of handling unknown messages, grouping into batches, and dispatching.

I guess you'd want a third function `processSingleMessage` for the `else` case below...?  Maybe there's an even clearer way to refactor it that's not jumping out at me.

As-is though this function is doing a lot of different things and would be nice to refactor somewhat.

Reply:
I agree that this code could do with some refactoring. However, I would prefer to do that in a separate PR because it will make the diff much harder to review. There are other things here like `processRuntimeMessage` and `observeNonRuntimeMessage` having lot of duplicate code that could be cleaned.

Reply:
I refactored this - consolidated all non-empty batch processing into a `processBatch` function. That seemed cleaner to me than processing modern vs non-modern messages. Take a look.